applications/ray/kuberay-tpu-webhook/main.go

```diff
@@ -251,85 +284,72 @@ func validateRayCluster(admissionReview *admissionv1.AdmissionReview) (*admissio
-		}
-		if containerRequestingTPUs(containers...) {
-			topology := workerGroupSpec.Template.Spec.NodeSelector["cloud.google.com/gke-tpu-topology"]
-			isMultiHost, err := isTPUMultiHost(topology)
-			if err != nil {
-				return nil, err
-			}
-			if isMultiHost {
-				joinedHostNames, err := genDNSHostnames(workerGroupSpec)
-				if err != nil {
-					return nil, err
-				}
-				injectHostnames(joinedHostNames, workerGroupSpec, i, &patches)
+func getEnvironmentVariable(varName string, container corev1.Container) string {
+	if container.Env != nil && len(container.Env) > 0 {
+		for _, envVar := range container.Env {
+			if envVar.Name == varName {
+				return envVar.Value
 			}
 		}
 	}
-
-	patchBytes, err := json.Marshal(patches)
-	if err != nil {
-		return nil, err
-	}
-
-	// Create AdmissionResponse
-	admissionResponse := &admissionv1.AdmissionResponse{
-		UID:     admissionReview.Request.UID,
-		Allowed: true,
-		Patch:   patchBytes,
-		PatchType: func() *admissionv1.PatchType {
-			pt := admissionv1.PatchTypeJSONPatch
-			return &pt
-		}(),
-	}
-	return admissionResponse, nil
+	return ""
 }
 
-func hasWorkerID(container corev1.Container) bool {
-	if container.Env != nil && len(container.Env) > 0 {
-		for _, envVar := range container.Env {
-			if envVar.Name == "TPU_WORKER_ID" {
-				return true
+// get next replica ID to assign a pod to
+func getReplicaIndex() int {
+	if sliceToWorkers == nil {
+		return 0
+	}
+	next_lowest_id := math.MaxInt32
+	for slice, workerList := range sliceToWorkers {
+		runningPods := 0
+		for _, worker := range workerList {
+			if worker.isRunning {
+				runningPods++
+			}
+		}
+		if runningPods < int(slice.numOfHosts) {
+			if slice.replicaIndex < next_lowest_id {
+				next_lowest_id = slice.replicaIndex
 			}
 		}
 	}
-	return false
+	return next_lowest_id
 }
 
-func hasWorkerName(container corev1.Container) bool {
-	if container.Env != nil && len(container.Env) > 0 {
-		for _, envVar := range container.Env {
-			if envVar.Name == "TPU_NAME" {
-				return true
+// returns next lowest TPU_WORKER_ID in pod slice and updates mappings
+func getNextWorkerID(podSlice slice, replicaIndex int) int {
+	tpuWorkerID := 0
+	if sliceToWorkers[podSlice] == nil {
+		newWorker := worker{tpuWorkerID, replicaIndex, true}
+		sliceToWorkers[podSlice] = []worker{newWorker}
+	} else {
+		nextLowestID := math.MaxInt32
+		replacePod := false
+		// iterate through existing workers and check if any have been deleted
+		for _, worker := range sliceToWorkers[podSlice] {
+			if worker.isRunning == false && worker.workerIndex < nextLowestID {
+				replacePod = true
+				nextLowestID = worker.workerIndex
+			}
+		}
+		// reassign next lowest TPU_WORKER_ID if pod has been deleted
+		if replacePod == true {
+			for _, worker := range sliceToWorkers[podSlice] {
+				// set worker.isRunning to true now that pod is being re-created
+				if worker.workerIndex == nextLowestID {
+					worker.isRunning = true
```

Comment:
What if this pod creation failed?

Reply:
We sort of just assume that if it gets to this point the pod creation will succeed, the admission reponse is always set with `Allowed: true` and it's only possible for it to fail if the json.Marshal fails which should never happen.